### DISCUSSION TITLE
Integrating Appium and Code magic

### DISCUSSION THREAD

**Discussion by Anandhakrishnansm** [codemagic-ci-cd/codemagic-docs#1761] [Q&A] [answered]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

React Native

### Steps to reproduce

My requirement is without using any kind of cloud-based mobile device or a real device , can we run our Appium script only with using code magic.

ie; headless mode without using real/cloud based mobile devices.

### Expected results

I want to run my wdio.conf.js file automatically when the build generation completed in code magic.
Please note that , I don't have a real/cloud based mobile devices.

### Actual results

No Result

**Comment by himesh-cm** [comment, marked as answer]
Hi @Anandhakrishnansm, You can use Appium and inbuilt emulators on build machines to run the tests; please check this [doc](https://docs.codemagic.io/yaml-testing/testing/#react-native-integration-tests-using-appium-and-emulator) which serves a good starting point.

Let us know if you hit any roadblocks.

**Reply by Anandhakrishnansm** [reply]
`exports.config = {
  services: ['appium'],
  port: 4723,
  runner: 'local',
  specs: [
    './tests/specs/**/*.js'
  ],
  capabilities: [{
     maxInstances: 1,
     browserName: '',
     appiumVersion: '1.13.0',
     platformName: 'Android',
     platformVersion: '<emulator platform version>',    // Specify your emulator details
     deviceName: '<emulator name>',
     app: '<path to APK>',
     automationName: 'UiAutomator2'
  }],

  logLevel: 'trace',
  bail: 0,
  waitforTimeout: 10000,
  connectionRetryTimeout: 90000,
  connectionRetryCount: 3,
  framework: 'mocha',
  reporters: ['spec'],
  mochaOpts: {
    ui: 'bdd'
    timeout: 60000
  }
}`

Usually in app: <path to APK>, I will be giving my .apk path which is stored in my machine. But when it comes to the codeMagic, we need to give the latest generated .apk file path instead right?

How can we do that?

**Reply by Anandhakrishnansm** [reply]
Also, I need to keep my Appium automation code in the same repository Which is used by the Mobile developers. Right??

**Comment by Anandhakrishnansm** [comment]
Hi @himesh-cm 

"You can use Appium and inbuilt emulators on build machines to run the tests" - Build machines means- is it my machine or the codeMagic virtual machine ??

because its's says :- **"In the context of CodeMagic, the build machines do not provide inbuilt emulators for running tests with Appium. CodeMagic primarily focuses on building and deploying mobile applications rather than providing inbuilt emulators or simulators for testing purposes."**

From my understanding, We need a target device(virtual or real) to run our Appium automation code.

It would be great if you can confirm this.

Regards,
Anandha

**Comment by Anandhakrishnansm** [comment]
@himesh-cm Please find the diagram given.
**I can explain my requirement in detail.**

1. There is a common repository called Mobile Repo.
In this repo, we will be having our mobile app development code and appium automation script and this repo is connected with codeMagic.
2. While raising a PR , CodeMagic will create a built-in virtual/build machine and it will start generating the build.
3. I want to run my appium script in the same built-in virtual/build machine once after the build generation completed.- **This is my requirement** 

**Please note : I don't want to use a real/Cloud based mobile devices to run my appium script. I want to run it in the same virtual machine created by codemagic for the build generation**

Is this requirement possible?


![image (51)](https://github.com/codemagic-ci-cd/codemagic-docs/assets/102238148/4441929e-0d5a-47e1-b8b6-eb295b76e8bb)

**Reply by himesh-cm** [reply]
Hi @Anandhakrishnansm, Apologies for the delayed reply; yes it is possible as per the flow shared by you. I suggest please create a new app first and set up the `codemagic.yaml` file which defines all the necessary steps as per the doc shared earlier, Please let us know how it goes.

**Comment by himesh-cm** [comment]
Hi @Anandhakrishnansm , While I’m closing our discussion for now, if you have any other questions or suggestions, feel free to reply back or start a new discussion anytime. Have a nice day!
